include/soci/type-holder.h

```diff
@@ -44,49 +50,310 @@ T* checked_ptr_cast(U* ptr)
+        case db_int64:
+            delete val_.i64;
+            break;
+        case db_uint8:
+            delete val_.u8;
+            break;
+        case db_uint16:
+            delete val_.u16;
+            break;
+        case db_uint32:
+            delete val_.u32;
+            break;
+        case db_uint64:
+            delete val_.u64;
+            break;
+        case db_date:
+            delete val_.t;
+            break;
+        case db_blob:
+        case db_xml:
+        case db_string:
+            delete val_.s;
+            break;
+        default:
+            break;
+        }
+    }
+
+    template <typename T>
+    T get()
+    {
+        switch (dt_)
+        {
+        case db_int8:
+            return soci_cast<T, int8_t>::cast(*val_.i8);
+        case db_int16:
+            return soci_cast<T, int16_t>::cast(*val_.i16);
+        case db_int32:
+            return soci_cast<T, int32_t>::cast(*val_.i32);
+        case db_int64:
+            return soci_cast<T, int64_t>::cast(*val_.i64);
+        case db_uint8:
+            return soci_cast<T, uint8_t>::cast(*val_.u8);
+        case db_uint16:
+            return soci_cast<T, uint16_t>::cast(*val_.u16);
+        case db_uint32:
+            return soci_cast<T, uint32_t>::cast(*val_.u32);
+        case db_uint64:
+            return soci_cast<T, uint64_t>::cast(*val_.u64);
+        case db_double:
+            return soci_cast<T, double>::cast(*val_.d);
+        case db_date:
+            return soci_cast<T, std::tm>::cast(*val_.t);
+        case db_blob:
+        case db_xml:
+        case db_string:
+            return soci_cast<T, std::string>::cast(*val_.s);
+        default:
+            throw std::bad_cast();
+        }
+    }
 
 private:
-    T * t_;
+    holder(db_type dt, void* val) : dt_(dt)
+    {
+        switch (dt_)
+        {
+        case db_double:
+            val_.d = static_cast<double*>(val);
+            break;
+        case db_int8:
+            val_.i8 = static_cast<int8_t*>(val);
+            break;
+        case db_int16:
+            val_.i16 = static_cast<int16_t*>(val);
+            break;
+        case db_int32:
+            val_.i32 = static_cast<int32_t*>(val);
+            break;
+        case db_int64:
+            val_.i64 = static_cast<int64_t*>(val);
+            break;
+        case db_uint8:
+            val_.u8 = static_cast<uint8_t*>(val);
+            break;
+        case db_uint16:
+            val_.u16 = static_cast<uint16_t*>(val);
+            break;
+        case db_uint32:
+            val_.u32 = static_cast<uint32_t*>(val);
+            break;
+        case db_uint64:
+            val_.u64 = static_cast<uint64_t*>(val);
+            break;
+        case db_date:
+            val_.t = static_cast<std::tm*>(val);
+            break;
+        case db_blob:
+        case db_xml:
+        case db_string:
+            val_.s = static_cast<std::string*>(val);
+            break;
+        default:
+            break;
```

Comment:
I don't see why this should throw as the constructor is private and it is only called during preparation for statement execution.

Reply:
This should be unreachable, right? I.e. it might be better to use `std::unreachable()` here but as long as we can't do this, throwing seems better than doing nothing.